File: afflyasnavigation/src/main/java/com/afflyas/afflyasnavigation/notification/ANNotification.java
package com.afflyas.afflyasnavigation.notification;

import android.os.Parcel;
import android.os.Parcelable;
import android.text.TextUtils;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.ColorInt;
import androidx.annotation.Nullable;

public class ANNotification implements Parcelable {

    @Nullable
    private String text; // can be null, so notification will not be shown

    @ColorInt
    private int textColor; // if 0 then use default value

    @ColorInt
    private int backgroundColor; // if 0 then use default value

    public ANNotification() {
        // empty
    }

    private ANNotification(Parcel in) {
        text = in.readString();
        textColor = in.readInt();
        backgroundColor = in.readInt();
    }

    public boolean isEmpty() {
        return TextUtils.isEmpty(text);
    }

    @Nullable
    public String getText() {
        return text;
    }

    public int getTextColor() {
        return textColor;
    }

    public int getBackgroundColor() {
        return backgroundColor;
    }

    public static ANNotification justText(String text) {
        return new Builder().setText(text).build();
    }

    public static List<ANNotification> generateEmptyList(int size) {
        List<ANNotification> notificationList = new ArrayList<>();
        for (int i = 0; i < size; i++) {
            notificationList.add(new ANNotification());
        }
        return notificationList;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(text);
        dest.writeInt(textColor);
        dest.writeInt(backgroundColor);
    }

    public static class Builder {
        @Nullable
        private String text;
        @ColorInt
        private int textColor;
        @ColorInt
        private int backgroundColor;

        public Builder setText(String text) {
            this.text = text;
            return this;
        }

        public Builder setTextColor(@ColorInt int textColor) {
            this.textColor = textColor;
            return this;
        }

        public Builder setBackgroundColor(@ColorInt int backgroundColor) {
            this.backgroundColor = backgroundColor;
            return this;
        }

        public ANNotification build() {
            ANNotification notification = new ANNotification();
            notification.text = text;
            notification.textColor = textColor;
            notification.backgroundColor = backgroundColor;
            return notification;
        }
    }

    public static final Creator<ANNotification> CREATOR = new Creator<ANNotification>() {
        @Override
        public ANNotification createFromParcel(Parcel in) {
            return new ANNotification(in);
        }

        @Override
        public ANNotification[] newArray(int size) {
            return new ANNotification[size];
        }
    };

}

File: afflyasnavigation-kotlin/src/main/java/com/afflyas/afflyasnavigation/notification/ANNotification.kt
package com.afflyas.afflyasnavigation.notification

import android.os.Parcel
import android.os.Parcelable
import android.support.annotation.ColorInt
import android.text.TextUtils
import java.util.ArrayList

class ANNotification : Parcelable {

    constructor() {
        // empty
    }

    var text: String? = null
        private set // can be null, so notification will not be shown

    @ColorInt
    var textColor: Int = 0
        private set // if 0 then use default value

    @ColorInt
    var backgroundColor: Int = 0
        private set // if 0 then use default value

    val isEmpty: Boolean
        get() = TextUtils.isEmpty(text)

    private constructor(inputParcel: Parcel) {
        text = inputParcel.readString()
        textColor = inputParcel.readInt()
        backgroundColor = inputParcel.readInt()
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(text)
        dest.writeInt(textColor)
        dest.writeInt(backgroundColor)
    }

    companion object {

        fun justText(text: String): ANNotification {
            return Builder().setText(text).build()
        }

        fun generateEmptyList(size: Int): ArrayList<ANNotification> {
            val notificationList = ArrayList<ANNotification>()
            for (i in 0 until size) {
                notificationList.add(ANNotification())
            }
            return notificationList
        }

        val CREATOR: Parcelable.Creator<ANNotification> = object : Parcelable.Creator<ANNotification> {
            override fun createFromParcel(inputParcel: Parcel): ANNotification {
                return ANNotification(inputParcel)
            }

            override fun newArray(size: Int): Array<ANNotification?> {
                return arrayOfNulls(size)
            }
        }
    }

    class Builder {
        private var text: String? = null
        @ColorInt
        private var textColor: Int = 0
        @ColorInt
        private var backgroundColor: Int = 0

        fun setText(text: String?): Builder {
            this.text = text
            return this
        }

        fun setTextColor(@ColorInt textColor: Int): Builder {
            this.textColor = textColor
            return this
        }

        fun setBackgroundColor(@ColorInt backgroundColor: Int): Builder {
            this.backgroundColor = backgroundColor
            return this
        }

        fun build(): ANNotification {
            val notification = ANNotification()
            notification.text = text
            notification.textColor = textColor
            notification.backgroundColor = backgroundColor
            return notification
        }
    }


}